### ISSUE TITLE
"training" argument change after model is defined

### ISSUE THREAD

**Issue by palatos** [tensorflow/tensorflow#44465]
**System information**
- TensorFlow version (you are using): 2.3.0

Say I define two models.

model1

```
input_shape = (224, 224, 3)
inputs = keras.Input(shape=input_shape)
x = tf.keras.layers.experimental.preprocessing.RandomRotation(0.5)(inputs, training = True)
x = tf.keras.layers.experimental.preprocessing.Rescaling(1./255.)(x)
x = tf.keras.layers.Convolution2D(32,(3,3))(x)
x = tf.keras.layers.MaxPool2D((2,2))(x)
x = tf.keras.layers.Flatten()(x)
out = tf.keras.layers.Dense(2,activation = 'softmax')(x)
model1 = keras.Model(inputs, out)
```
model2
```
input_shape = (224, 224, 3)
inputs = keras.Input(shape=input_shape)
x = tf.keras.layers.experimental.preprocessing.RandomRotation(0.5)(inputs, training = False)
x = tf.keras.layers.experimental.preprocessing.Rescaling(1./255.)(x)
x = tf.keras.layers.Convolution2D(32,(3,3))(x)
x = tf.keras.layers.MaxPool2D((2,2))(x)
x = tf.keras.layers.Flatten()(x)
out = tf.keras.layers.Dense(2,activation = 'softmax')(x)
model2= keras.Model(inputs, out)
```

Where the only difference is that in model2 the preprocessing augmentation layer has "training" set to False. There have been many instances where I simply want to change this parameter to False/True, after the model was defined, and I can find no way to do this besides creating a new model. First time I encountered this issue was in fine tuning, but more recently when trying to generate gradcam explanations. I imagine this has countless other occurrences, and might also happen with layers like BatchNorm.

Is there already a way to switch the training state of these types of layers? (Note I don't mean the layer.trainable attribute).
If not, could this be implemented? It would be very useful.

**Comment by bhack** [comment]
@palatos Do want to have a different behavior than `K.learning_phase()`?

**Comment by palatos** [comment]
@bhack I think if I use set_learning_phase(1) for instance, model2 would still have 'training = False' (I just ran a test and it seems to indicate this). Is there any way to change model2's 'training' parameter to 'True' after the model was defined?

**Comment by bhack** [comment]
I meant that if you don't set the argument the `call` will be in sync with the learning phase 
https://github.com/tensorflow/tensorflow/blob/a87725285625c151055aaf7e8bab10d0c50f410a/tensorflow/python/keras/layers/preprocessing/image_preprocessing.py#L228-L229

**Comment by palatos** [comment]
My problem is I'm using a model pretrained by someone else, and I believe they had set this parameter when generating the model. Once I load their model can I override whatever they set for training by using this learning_phase idea?

**Comment by bhack** [comment]
> Once I load their model can I override whatever they set for training by using this learning_phase idea?
In your case I don't think you will override with  learning_phase.

**Comment by palatos** [comment]
Yeah I did a mock test here and it seems like indeed I can't. If someone set training = 'False' and saved the model it seems like that's it and it cannot be changed anymore.
What I mean is, would it be hard to implement it so that this can be as easily changed as say, layer.trainable? Would that break something major?

**Comment by bhack** [comment]
Do you mean something similar to https://github.com/tensorflow/tensorflow/issues/39036?

**Comment by palatos** [comment]
Indeed, that user seems to have the same need as me. The solution he proposes would fix my problem too.

**Comment by bhack** [comment]
I think you can close this ticket and upvote that one.

**Comment by palatos** [comment]
Ok, will do.

**Comment by zachcoleman** [comment]
@bhack have you tried saving the weights instead of the Model object? 

Like this example (I changed it to BatchNormalization cause of different version of TF but I assume it still applies):

```python
import tensorflow as tf
import numpy as np

input_shape = (224, 224, 3)
inputs = tf.keras.Input(shape=input_shape)
x = tf.keras.layers.BatchNormalization()(inputs, training=True)
x = tf.keras.layers.Convolution2D(32,(3,3))(x)
x = tf.keras.layers.MaxPool2D((2,2))(x)
x = tf.keras.layers.Flatten()(x)
out = tf.keras.layers.Dense(2,activation = 'softmax')(x)
model1 = tf.keras.Model(inputs, out)

input_shape = (224, 224, 3)
inputs = tf.keras.Input(shape=input_shape)
x = tf.keras.layers.BatchNormalization()(inputs, training=False)
x = tf.keras.layers.Convolution2D(32,(3,3))(x)
x = tf.keras.layers.MaxPool2D((2,2))(x)
x = tf.keras.layers.Flatten()(x)
out = tf.keras.layers.Dense(2,activation = 'softmax')(x)
model2 = tf.keras.Model(inputs, out)

# have different weights
[print(np.all((a==b)), a.shape, b.shape) for a, b in zip(model1.weights, model2.weights)]

model1.save_weights("temp.json")
model2.load_weights("temp.json")

# have the same weights
[print(np.all((a==b)), a.shape, b.shape) for a, b in zip(model1.weights, model2.weights)]

# the respective calls that will be invoked on the models
# model1.layers[1]._get_training_value(training=True)
# model2.layers[1]._get_training_value(training=False)
```

You will have your pre-trained weights in a different (but equivalent) model where you can control the __training__ parameter.
